numpy/core/tests/test_ufunc.py

```diff
@@ -597,6 +597,78 @@ def test_output_argument(self):
         umt.inner1d(a, b, out=c[..., 0])
         assert_array_equal(c[..., 0], np.sum(a*b, axis=-1), err_msg=msg)
 
+    def test_axes_argument(self):
+        # matrix_multiply signature: '(m,n),(n,p)->(m,p)'
+        a = np.arange(12).reshape((2, 3, 2))
+        b = np.arange(8).reshape((2, 2, 2, 1)) + 1
+        c = umt.matrix_multiply(a, b)
+        assert_array_equal(c, np.matmul(a, b))
+        # Default.
+        c = umt.matrix_multiply(a, b, axes=[(-2, -1), (-2, -1), (-2, -1)])
+        assert_array_equal(c, np.matmul(a, b))
+        # Default with explicit axes.
+        c = umt.matrix_multiply(a, b, axes=[(1, 2), (2, 3), (2, 3)])
+        assert_array_equal(c, np.matmul(a, b))
+        # swap some axes.
+        c = umt.matrix_multiply(a, b, axes=[(0, -1), (1, 2), (-2, -1)])
+        assert_array_equal(c, np.matmul(a.transpose(1, 0, 2),
+                                        b.transpose(0, 3, 1, 2)))
+        # Default with output array.
+        c = np.empty((2, 2, 3, 1))
+        d = umt.matrix_multiply(a, b, out=c, axes=[(1, 2), (2, 3), (2, 3)])
+        assert_(c is d)
+        assert_array_equal(c, np.matmul(a, b))
+        # Transposed output array
+        c = np.empty((1, 2, 2, 3))
+        d = umt.matrix_multiply(a, b, out=c, axes=[(-2, -1), (-2, -1), (3, 0)])
+        assert_(c is d)
+        assert_array_equal(c, np.matmul(a, b).transpose(3, 0, 1, 2))
+        # Check errors for inproperly constructed axes arguments.
+        assert_raises(TypeError, umt.matrix_multiply, a, b, axis=1)
+        assert_raises(TypeError, umt.matrix_multiply, a, b, axes=1)
+        assert_raises(TypeError, umt.matrix_multiply, a, b,
+                      axes=((-2, -1), (-2, -1), (-2, -1)))
+        assert_raises(ValueError, umt.matrix_multiply, a, b, axes=[])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[-1, -1, -1])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[(-2, -1), (-2, -1), -1])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[[-2, -1], [-2, -1], [-2, -1]])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[(-2, -1), (-2, -1), [-2, -1]])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[(-2, -1), (-2, -1), None])
+        assert_raises(ValueError, umt.matrix_multiply, a, b,
+                      axes=[(-2, -1)])
```

Comment:
Nit: I find this kind of thing reads better as:
```
assert_raises(ValueError,
    the_function, and, its, arguments)
```

Reply:
OK, I solved this by setting `mm = umt.matrix_multiply` so that everything fits on a single line.